Allow factory system descriptor to initialize from multiple cable descriptors

jpanasiukTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/33818

### Problem description

It would be useful for large-scale systems (e.g., BH Exabox with 8x16 meshes) to have their cabling descriptor split across multiple descriptor files for easier management - intrapod cables, interpod cables, inter-superpod cables, etc. Previously, `CablingGenerator` only supported single descriptor files, making it difficult to maintain and validate these distributed cabling configurations.

This feature will be used as part of superpod work where we expect to have at least 2 different files (16 devices + 20 devices).

### What's changed

**Core functionality:**
- `CablingGenerator` constructor now accepts directories containing multiple `.textproto` files, automatically discovering and merging them
- Added merge validation: detects port conflicts (same port → different destinations) across descriptors and deduplicates identical connections
- Added torus-specific merging: X_TORUS + Y_TORUS → XY_TORUS by combining their `inter_board_connections` (different architectures correctly rejected)
- Graph templates validated for structural consistency (motherboard, boards, node types) during merge

**Merge support & limitations:**

**Supported:**
- Merging graph-level `internal_connections` from multiple descriptors
- Torus-specific merging: X_TORUS + Y_TORUS → XY_TORUS (same architecture only)
- Cross-descriptor connections on different ports (port exhaustion validated at FSD generation)
- Automatic deduplication of identical connections
- Graph templates must have **identical children** across all descriptor files (same nodes and subgraphs)

**Not supported:**
- **Graph templates with different children** (e.g., File1 has {node1, node2}, File2 has {node2, node3})
- Merging nodes with different templates (different motherboards or board structures)
- Merging nodes with different architectures (e.g., WH_GALAXY + BH_GALAXY)
- Merging non-torus node-level `inter_board_connections` (must match exactly across descriptors)
- Same port connecting to different destinations within a single descriptor (conflict)
- Cross-architecture torus merging (e.g., WH_X_TORUS + BH_X_TORUS)

**Testing & validation:**
- Added 26 comprehensive tests organized into 6 categories: Sanity (2), Torus Merge (4), Split/Merge (3), Negative (11), Operator== (5), Descriptor Loading (1)
- Added `operator==` for `CablingGenerator` and helper comparison functions to validate merge correctness
- Split/merge round-trip validation for 8x16 WH XY_TORUS superpod (2-16 way splits), 5-node Y_TORUS, and 16-node N300 clusters

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/21132784012
- [ ] New/Existing tests provide coverage for changes

tt-asaigal commented on tools/scaleout/cabling_generator/cabling_generator.cpp:
```diff
@@ -320,77 +320,498 @@ void populate_deployment_hosts_from_hostnames(
-    // Populate the host_id_to_node_ map
+void CablingGenerator::merge(const CablingGenerator& other, const std::string& source_file) {
+    // Validate node_templates_ are identical (must match exactly)
+    for (const auto& [name, other_template] : other.node_templates_) {
```
This logic should be in a validation fn

tt-asaigal commented on tools/scaleout/cabling_generator/cabling_generator.cpp:
```diff
@@ -320,77 +364,713 @@ void populate_deployment_hosts_from_hostnames(
+                }
+            }
+        } else {
+            // New template - add it
```
should we assert here?

jpanasiukTT committed:
Add core descriptor merging infrastructure

- Add DescriptorMerger class for combining multiple cabling descriptors
- Support merging cluster descriptors from multiple .textproto files
- Add comprehensive test suite for descriptor merging
- Extract load_descriptor_from_textproto to common protobuf_utils.hpp
- Remove unused vector-based constructors
- Remove unused save_factory_system_descriptor wrapper
- Clean up and refactor merge validation logic

jpanasiukTT committed:
Add torus-specific merging support

- Implement torus-compatible node template merging
- Add logic to merge X_TORUS + Y_TORUS into XY_TORUS
- Add comprehensive torus merge testing
- Refactor to eliminate torus connection duplication via node.cpp helpers
- Support merging inter_board_connections for torus nodes
- Add validation for architecture compatibility in torus merges

jpanasiukTT committed:
Integrate merge functionality into CablingGenerator

- Move merge logic from DescriptorMerger into CablingGenerator class
- Add directory processing support in CablingGenerator constructor
- Remove standalone descriptor_merger.cpp/.hpp files
- Integrate merge with build_graph_instance and node creation
- Add support for merging graph-level internal_connections
- Enhance test coverage for split/merge roundtrip validation
- Add operator== for CablingGenerator equality comparison

jpanasiukTT committed:
Code review improvements and refactoring

- Add Architecture enum and virtual get_architecture() to NodeBase hierarchy
- Remove protobuf_utils.hpp and integrate into cabling_generator
- Consolidate load_descriptor template functions
- Add is_torus() helper function to node.cpp
- Improve error messages with file source context
- Add validation for protobuf map lookups (prevent FatalException)
- Refactor node template validation and merging logic
- Remove unused helper functions and improve code organization
- Update README with current implementation details

jpanasiukTT committed:
Test fixes and code review improvements

- Improve test assertions to use operator== instead of EXPECT_NO_THROW
- Make CablingGenerator::merge() and find_descriptor_files() private
- Add equality validation for cross-descriptor connections test
- Add reflexive equality checks for LoadAllAvailableDescriptors test
- Move utility functions to private section for better encapsulation

jpanasiukTT committed:
pr review, reducing code redundancy plus some tests grouping

jpanasiukTT commented on tools/scaleout/cabling_generator/cabling_generator.cpp:
```diff
@@ -320,82 +712,375 @@ void populate_deployment_hosts_from_hostnames(
+            cluster_descriptor.root_instance(), cluster_descriptor, &deployment_descriptor, "", node_templates_);
+        validate_host_id_uniqueness();
+        populate_host_id_to_node();
+        generate_logical_chip_connections();
```
Here we walk `ResolvedGraphInstance` 3 separate times, it shouldn't be that big of a performance difference and it should make a little easier to understand but I would like to get feedback on that.

jpanasiukTT committed:
restrict merging nodes and subgraphs

tt-asaigal commented on tools/scaleout/cabling_generator/cabling_generator.cpp:
```diff
@@ -770,6 +1642,15 @@ CableLength calc_cable_length(
     }
 
+    // Constants for Galaxy node physical dimensions
+    constexpr double TRAY_HEIGHT_U = 1.25;         // U per tray in Galaxy nodes
```
Looks like a merge conflict

jpanasiukTT committed:
Add descriptor merger tests and optimize test performance

- Add test_descriptor_merger to CI test suite (run_tools_tests.sh)
- Optimize test runtime by testing only powers of 2 for split counts (2, 4, 8, 16)
  instead of all values (2-16), reducing test time from ~22s to ~1s
- Suppress verbose logging during tests with TT_METAL_LOGGER_LEVEL=Fatal
- Fix post-rebase issues: reinstate removed tests and restore missing constructors
- Add support for non-path cluster descriptors
- Code review improvements: reduce redundancy, add clarifying comments

jpanasiukTT committed:
Address Clang-Tidy warnings from code review

- Remove redundant 'static' keyword from build_graph_instance wrappers in anonymous namespace
- Replace .count() with .contains() for more readable membership checks (C++20)
- All 15 Clang-Tidy warnings resolved

jpanasiukTT committed:
Fix remaining Clang-Tidy warnings

- Remove redundant 'static' from 10 more functions in anonymous namespace
- Replace .count() with .contains() in 10 more locations for readability
- All Clang-Tidy warnings now resolved (30 total fixes)

jpanasiukTT commented:
/codeowners ping

agupta-TT commented:
For the test would it be possible to get sample usage of the descriptor merge under the tools/tests/scaleout/cabling_descriptors directory

jpanasiukTT committed:
Fix duplicate build_graph_instance definition after rebase

jpanasiukTT committed:
Add tests that use files from cabling_descriptors directory

jpanasiukTT committed:
copilot

jpanasiukTT committed:
clang tidy issues

jpanasiukTT committed:
Merge branch 'main' into jpanasiuk/simple-desc-merge

jpanasiukTT committed:
Merge branch 'main' into jpanasiuk/simple-desc-merge

jpanasiukTT committed:
Fix remaining Clang-Tidy warnings

- Remove redundant 'static' from build_normalized_graph_connection_set
- Replace .count() with .contains() in 2 more locations
- Remove 'else' after 'throw' statement

jpanasiukTT committed:
add contains
